### PR TITLE
Pull env variables from EB web server instead of ecs task definition

### PR DESCRIPTION
**Pull request by jeet4320**
Closes https://www.wrike.com/open.htm?id=262256123

## What

1. [eb printenv env_name](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/eb3-printenv.html) pulls env variables from EB env_name 
2. Pass these env variables while `docker build` and `docker run`. 
3. `Docker build` will use variables which are defined in `Dockerfile` like [snaptx](https://github.com/bodyshopbidsdotcom/snapsheet-tx/pull/1029) dockerfile is using NPM_TOKEN as a variable to do `npm install`. 
4. This PR won't change any environment variables and it will just utilize them during `docker build` and `docker run`

## Why
While doing `docker build` NPM_TOKEN is needed to build some projects. If we add NPM_TOKEN as env variable to EB env then it is not available to pass NPM_TOKEN using current method which pulls env variables from task definitions. Task definitions get created after first deployment but before this we need NPM_TOKEN to build project.

### CODE REVIEW

File: lib/push_eb.py [hunk 1]
```diff
@@ -67,7 +67,7 @@ def dockerrun_config(self, action, command):
 
   def container_definitions(self, action, command):
     use_ports = [{"hostPort": 80, "containerPort": 80}] if 'web' in action else []
-    config = {"image": self.image_uri, "name": action, "command": command.strip().split(), "portMappings": use_ports}
+    config = {"image": self.image_uri, "name": action, "command": [ "sh", "-c", command.strip() ], "portMappings": use_ports}
```
Comment (jeet4320) [inline lib/push_eb.py]: this is needed to make env variable work like $PORT in [Procfile](https://github.com/bodyshopbidsdotcom/snapsheet-tx/blob/release/Procfile)

File: lib/aws/eb.py [hunk 2]
```diff
@@ -89,7 +90,16 @@ def describe_environments(self, app_name, env_name):
       self._current_environments = new_envs
     return self._current_environments
 
-  def get_environment_variables(self, action):
+  # get environment variables from EB web environment
+  def get_eb_environment_variables(self, action):
```
Comment (whatisdot) [inline lib/aws/eb.py]: Yesterday we discovered the api that `ebcli` uses to get/set environment variables. We can use this, and remove the need for multiple `get_*_environment_variables` functions.

File: lib/aws/eb.py [hunk 3]
```diff
@@ -91,8 +91,13 @@ def describe_environments(self, app_name, env_name):
     return filtered_envs
 
   def get_environment_variables(self, action):
-    response = ECS.describe_task_definition(self.cluster_name(action))
-    return response['taskDefinition']['containerDefinitions'][0]['environment']
+    response = self.describe_configuration_settings(self.project_name(), EB.environment_name(self.project_name(), self.args.environment, action))
+    env_vars_hash = {}
+    for optionsetting in response['ConfigurationSettings'][0]['OptionSettings']:
```
Comment (whatisdot) [inline lib/aws/eb.py]: Use kindly_find here

File: lib/aws/eb.py [hunk 4]
```diff
@@ -101,6 +106,14 @@ def cluster_name(self, action):
       self.get_environment_id(action)[2:]
     ])
 
+  def describe_configuration_settings(self, app_name, env_name):
```
Comment (whatisdot) [inline lib/aws/eb.py]: This should be a class method.

File: lib/push_eb.py [hunk 5]
```diff
@@ -134,22 +134,25 @@ def command(self):
     self.logger.debug('New Version: '+self.version)
     self.config = ConfigManager(self.logger, self.region(), self.ecr_image_uri(), self.project_name(), self.args.environment)
     try:
-      self.build()
+      env_vars = self.get_environment_variables('web')
```
Comment (whatisdot) [inline lib/push_eb.py]: These should go in the build function.

Comment (jeet4320) [reply lib/push_eb.py]: yes its a parameter in `self.build` and it needs code at line 138 before passing env_vars

Comment (jeet4320) [reply lib/push_eb.py]: I see what you mean

File: lib/aws/eb.py [hunk 6]
```diff
@@ -91,8 +91,13 @@ def describe_environments(self, app_name, env_name):
     return filtered_envs
 
   def get_environment_variables(self, action):
-    response = ECS.describe_task_definition(self.cluster_name(action))
-    return response['taskDefinition']['containerDefinitions'][0]['environment']
+    response = EB.describe_configuration_settings(self.project_name(), EB.environment_name(self.project_name(), self.args.environment, action))
+    env_vars_hash = {}
+    for optionsetting in response['ConfigurationSettings'][0]['OptionSettings']:
```
Comment (whatisdot) [inline lib/aws/eb.py]: (comment got removed, so clarifying)
Try this:
```python
for optionsetting in self.kindly_find(response, ['ConfigurationSettings', 0, 'OptionSettings']):
```

**Pull request merged**